### PR TITLE
Allow cache location configuration

### PR DESCRIPTION
**Pull request by haikoschol**
This is a first attempt to address #357 and #685.

I tried to move around as few things as possible, but I'm still not totally happy with the result. IMHO The most problematic part are import side effects, especially the creation of directories.

Instead of the environment variable `SCANCODE_TEMP_DIR` the code looks for `SCANCODE_TMP`, because that was already in use in `commoncode/fileutils.py`.

### CODE REVIEW

**Comment by pombredanne** [comment]
@haikoschol Thank you!

File: src/commoncode/fileutils.py [hunk 1]
```diff
@@ -51,7 +51,10 @@
 import stat
 import sys
 import tempfile
-
+from functools import partial
+from hashlib import md5
+from os.path import abspath
+from os.path import dirname
```
Comment (pombredanne) [inline src/commoncode/fileutils.py]: Do you mind keeping the imports sorted in a given "block" (e.g. here the stdlib imports)

File: src/licensedcode/__init__.py [hunk 2]
```diff
@@ -27,25 +27,18 @@
...
 from os.path import dirname
 from os.path import abspath
-from os.path import getsize
-from os.path import getmtime
 from os.path import join
 from os.path import exists
 
-from commoncode import fileutils
-
-
```
Comment (pombredanne) [inline src/licensedcode/__init__.py]: I kinda like 2 empty lines after imports

File: src/commoncode/fileutils.py [hunk 3]
```diff
@@ -164,6 +170,64 @@ def get_temp_dir(base_dir, prefix=''):
...
+    if cache_root and not os.path.exists(cache_root):
+        raise OSError('{} is set to non-existent directory: {}'.format(TMP_DIR_ENV, cache_root))
+
+    if not cache_root:
+        if dev_mode:
+            cache_root = dirname(dirname(dirname(abspath(__file__))))
+        else:
+            cache_root = os.path.expanduser('~')
+
+    # tree_hash = tree_checksum()
```
Comment (pombredanne) [inline src/commoncode/fileutils.py]: Remove this

File: src/licensedcode/cache.py [hunk 4]
```diff
@@ -24,22 +24,17 @@
...
 from __future__ import absolute_import, print_function
 
-from functools import partial
-from hashlib import md5
 import os
 from os.path import exists
-from os.path import getmtime
-from os.path import getsize
 from os.path import join
+from functools import partial
```
Comment (pombredanne) [inline src/licensedcode/cache.py]: Please keep imports sorted here and elsewhere

File: src/licensedcode/cache.py [hunk 5]
```diff
@@ -48,45 +43,20 @@
...
-# If this file exists at the root, the cache is always checked for consistency
-DEV_MODE = os.path.exists(os.path.join(root_dir, 'SCANCODE_DEV_MODE'))
-
-
 def get_or_build_index_through_cache(
         check_consistency=DEV_MODE,
         return_index=True,
         # used for testing only
-        _tree_base_dir=src_dir,
+        _tree_base_dir=None,
```
Comment (pombredanne) [inline src/licensedcode/cache.py]: Why this change?

File: src/scancode/cache.py [hunk 6]
```diff
@@ -41,9 +41,7 @@
 from commoncode.fileutils import path_to_unicode
 from commoncode.system import on_linux
 from commoncode import timeutils
-
-from scancode import scans_cache_dir
-
+from licensedcode import DEV_MODE
```
Comment (pombredanne) [inline src/scancode/cache.py]: I think this should be across all scrancode, not just licensedcode

File: src/commoncode/fileutils.py [hunk 7]
```diff
@@ -164,6 +170,64 @@ def get_temp_dir(base_dir, prefix=''):
...
+    cache_root = os.environ.get(TMP_DIR_ENV)
+
+    if cache_root and not os.path.exists(cache_root):
+        raise OSError('{} is set to non-existent directory: {}'.format(TMP_DIR_ENV, cache_root))
+
+    if not cache_root:
+        if dev_mode:
+            cache_root = dirname(dirname(dirname(abspath(__file__))))
+        else:
+            cache_root = os.path.expanduser('~')
```
Comment (pombredanne) [inline src/commoncode/fileutils.py]: This may need an abspath too

File: src/commoncode/fileutils.py [hunk 8]
```diff
@@ -164,6 +170,64 @@ def get_temp_dir(base_dir, prefix=''):
...
+    cache_path = build_cache_path(dir_name, dev_mode=dev_mode)
+    create_dir(cache_path)
+    return cache_path
+
+
+def build_cache_path(dir_name, dev_mode=False):
+    cache_root = os.environ.get(TMP_DIR_ENV)
+
+    if cache_root and not os.path.exists(cache_root):
+        raise OSError('{} is set to non-existent directory: {}'.format(TMP_DIR_ENV, cache_root))
```
Comment (pombredanne) [inline src/commoncode/fileutils.py]: Why raise an error there? The directory should be created instead IMHO

File: src/commoncode/fileutils.py [hunk 9]
```diff
@@ -164,6 +170,64 @@ def get_temp_dir(base_dir, prefix=''):
...
+
+    if not cache_root:
+        if dev_mode:
+            cache_root = dirname(dirname(dirname(abspath(__file__))))
```
Comment (pombredanne) [inline src/commoncode/fileutils.py]: This is the root of scancode's code, right?

File: src/commoncode/fileutils.py (hunk 8 above)
Comment (haikoschol) [reply src/commoncode/fileutils.py]: Setting the environment variable requires user intervention (unlike automatically picking cache dirs based on install method). Therefore I'd argue the user wants to be notified if there is a typo in the path, for example.
Creating a directory in the wrong place and proceeding with the scan might lead to filling up the wrong file system and could be time-consuming to diagnose.

File: src/licensedcode/cache.py (hunk 5 above)
Comment (haikoschol) [reply src/licensedcode/cache.py]: I wanted to minimize the need to import global variables. It tricky to avoid circular imports and with lower-case names, it is easy to get them mixed up with local variables.

File: src/commoncode/fileutils.py (hunk 9 above)
Comment (haikoschol) [reply src/commoncode/fileutils.py]: Yes. I decided against using the global `src_code` to avoid to many imports across modules. But see my comment below regarding `DEV_MODE`.

File: src/scancode/cache.py (hunk 6 above)
Comment (haikoschol) [reply src/scancode/cache.py]: Agreed. I thought about what would be a better place for it, probably `commoncode`. But it seemed like moving a few of those globals would drag a whole lot of stuff with it, so I opted for keeping changes to a minimum.
I can try to move all globals to `commoncode/__init__.py` (and uppercase them while I'm at it). Should I?

**Comment by sschuberth** [comment]
@pombredanne Can we please get an updated review from you on this? This is blocking to resolve https://github.com/nexB/scancode-toolkit/issues/685 for us.

**Comment by pombredanne** [comment]
@sschuberth Looking at this now! Sorry for the delay

**Comment by pombredanne** [comment]
@haikoschol I would like to rework this on top of the new code in #885 based on the detailed description in https://github.com/nexB/scancode-toolkit/issues/685#issuecomment-358929114

**Comment by haikoschol** [comment]
As discussed offline, this has been superseded by #885.
